File: web/src/scriptFonts.js
export const LATIN = 'latin'

export const LANG_SCRIPT = {
  ar: 'arabic',
  bg: 'cyrillic',
  el: 'greek',
  he: 'hebrew',
  hi: 'devanagari',
  mr: 'devanagari',
  ja: 'japanese',
  ko: 'korean',
  ru: 'cyrillic',
  th: 'thai',
  uk: 'cyrillic',
  zh: 'chinese',
  'zh-Hant': 'chinese',
}

export function scriptForLang(lang) {
  return LANG_SCRIPT[lang] ?? LATIN
}

const SCRIPT_FONTS = {
  cyrillic: {
    anger: '"Russo One", sans-serif',
    joy: '"M PLUS Rounded 1c", sans-serif',
    play: '"Pacifico", cursive',
    calm: '"Comfortaa", sans-serif',
    tender: '"Marck Script", cursive',
    drive: '"Unbounded", sans-serif',
    sad: '"Cormorant Garamond", serif',
    anxiety: '"Amatic SC", cursive',
    reflective: '"PT Serif", serif',
  },
  greek: {
    anger: '"Tektur", sans-serif',
    joy: '"Arima", sans-serif',
    play: '"Mansalva", cursive',
    calm: '"Comfortaa", sans-serif',
    tender: '"Mynerve", cursive',
    drive: '"Syne", sans-serif',
    sad: '"Cardo", serif',
    anxiety: '"Fira Sans Extra Condensed", sans-serif',
    reflective: '"Noto Sans", sans-serif',
  },
  arabic: {
    anger: '"Blaka", cursive',
    joy: '"Marhey", sans-serif',
    play: '"Lemonada", sans-serif',
    calm: '"Mada", sans-serif',
    tender: '"Playpen Sans Arabic", cursive',
    drive: '"Alexandria", sans-serif',
    sad: '"Amiri", serif',
    anxiety: '"Jomhuria", cursive',
    reflective: '"Noto Sans Arabic", sans-serif',
  },
  hebrew: {
    anger: '"Suez One", serif',
    joy: '"Varela Round", sans-serif',
    play: '"Gveret Levin", cursive',
    calm: '"M PLUS Rounded 1c", sans-serif',
    tender: '"Secular One", sans-serif',
    drive: '"Miriam Libre", sans-serif',
    sad: '"Frank Ruhl Libre", serif',
    anxiety: '"Amatic SC", cursive',
    reflective: '"Heebo", sans-serif',
  },
  devanagari: {
    anger: '"Bakbak One", sans-serif',
    joy: '"Modak", cursive',
    play: '"Ranga", cursive',
    calm: '"Palanquin", sans-serif',
    tender: '"Kalam", cursive',
    drive: '"Rajdhani", sans-serif',
    sad: '"Tiro Devanagari Hindi", serif',
    anxiety: '"Khand", sans-serif',
    reflective: '"Noto Sans Devanagari", sans-serif',
  },
  thai: {
    anger: '"Srisakdi", sans-serif',
    joy: '"Chonburi", cursive',
    play: '"Itim", cursive',
    calm: '"Mitr", sans-serif',
    tender: '"Charmonman", cursive',
    drive: '"Chakra Petch", sans-serif',
    sad: '"Taviraj", serif',
    anxiety: '"Sriracha", cursive',
    reflective: '"Sarabun", sans-serif',
  },
  japanese: {
    anger: '"Dela Gothic One", sans-serif',
    joy: '"Kosugi Maru", sans-serif',
    play: '"Hachi Maru Pop", cursive',
    calm: '"Zen Maru Gothic", sans-serif',
    tender: '"Klee One", cursive',
    drive: '"RocknRoll One", sans-serif',
    sad: '"Shippori Mincho", serif',
    anxiety: '"DotGothic16", sans-serif',
    reflective: '"Noto Sans JP", sans-serif',
  },
  korean: {
    anger: '"Black Han Sans", sans-serif',
    joy: '"Jua", sans-serif',
    play: '"Gaegu", cursive',
    calm: '"Gowun Dodum", sans-serif',
    tender: '"Nanum Pen Script", cursive',
    drive: '"Do Hyeon", sans-serif',
    sad: '"Nanum Myeongjo", serif',
    anxiety: '"Single Day", cursive',
    reflective: '"Gothic A1", sans-serif',
  },
  chinese: {
    anger: '"WDXL Lubrifont SC", sans-serif',
    joy: '"ZCOOL KuaiLe", sans-serif',
    play: '"Liu Jian Mao Cao", cursive',
    calm: '"ZCOOL XiaoWei", sans-serif',
    tender: '"Zhi Mang Xing", cursive',
    drive: '"ZCOOL QingKe HuangYou", sans-serif',
    sad: '"Noto Serif SC", serif',
    anxiety: '"Long Cang", cursive',
    reflective: '"Noto Sans SC", sans-serif',
  },
}

const FALLBACK_FONT = '"Noto Sans", ui-sans-serif, system-ui, sans-serif'

export function fontForScript(script, cluster) {
  return SCRIPT_FONTS[script]?.[cluster] ?? FALLBACK_FONT
}

export const SCRIPT_FONT_QUERY = {
  cyrillic:
    'family=Russo+One&family=M+PLUS+Rounded+1c:wght@700&family=Pacifico&family=Comfortaa:wght@500&family=Marck+Script&family=Unbounded:wght@700&family=Cormorant+Garamond:ital,wght@1,500&family=Amatic+SC:wght@400&family=PT+Serif:ital,wght@0,400',
  greek:
    'family=Tektur:wght@700&family=Arima:wght@600&family=Mansalva&family=Comfortaa:wght@500&family=Mynerve&family=Syne:wght@700&family=Cardo:ital,wght@1,400&family=Fira+Sans+Extra+Condensed:wght@300&family=Noto+Sans:ital,wght@0,400',
  arabic:
    'family=Blaka&family=Marhey:wght@500&family=Lemonada:wght@600&family=Mada:wght@400&family=Playpen+Sans+Arabic:wght@400&family=Alexandria:wght@700&family=Amiri:ital,wght@1,400&family=Jomhuria&family=Noto+Sans+Arabic:wght@400',
  hebrew:
    'family=Suez+One&family=Varela+Round&family=Gveret+Levin&family=M+PLUS+Rounded+1c:wght@400&family=Secular+One&family=Miriam+Libre:wght@700&family=Frank+Ruhl+Libre:wght@300&family=Amatic+SC:wght@400&family=Heebo:wght@400',
  devanagari:
    'family=Bakbak+One&family=Modak&family=Ranga:wght@700&family=Palanquin:wght@400&family=Kalam:wght@400&family=Rajdhani:wght@600&family=Tiro+Devanagari+Hindi:ital,wght@1,400&family=Khand:wght@500&family=Noto+Sans+Devanagari:wght@400',
  thai: 'family=Srisakdi:wght@700&family=Chonburi&family=Itim&family=Mitr:wght@400&family=Charmonman:wght@400&family=Chakra+Petch:ital,wght@0,600&family=Taviraj:ital,wght@1,400&family=Sriracha&family=Sarabun:ital,wght@0,400',
  japanese:
    'family=Dela+Gothic+One&family=Kosugi+Maru&family=Hachi+Maru+Pop&family=Zen+Maru+Gothic:wght@400&family=Klee+One:wght@400&family=RocknRoll+One&family=Shippori+Mincho:wght@500&family=DotGothic16&family=Noto+Sans+JP:wght@400',
  korean:
    'family=Black+Han+Sans&family=Jua&family=Gaegu:wght@400&family=Gowun+Dodum&family=Nanum+Pen+Script&family=Do+Hyeon&family=Nanum+Myeongjo:wght@400&family=Single+Day&family=Gothic+A1:wght@400',
  chinese:
    'family=WDXL+Lubrifont+SC&family=ZCOOL+KuaiLe&family=Liu+Jian+Mao+Cao&family=ZCOOL+XiaoWei&family=Zhi+Mang+Xing&family=ZCOOL+QingKe+HuangYou&family=Noto+Serif+SC:wght@300&family=Long+Cang&family=Noto+Sans+SC:wght@400',
}

const loadedScripts = new Set()

export function ensureScriptFontsLoaded(script) {
  if (script === LATIN || loadedScripts.has(script) || typeof document === 'undefined') return
  const query = SCRIPT_FONT_QUERY[script]
  if (!query) return
  loadedScripts.add(script)
  const link = document.createElement('link')
  link.rel = 'stylesheet'
  link.href = `https://fonts.googleapis.com/css2?${query}&display=swap`
  document.head.appendChild(link)
}

File: android/app/src/main/java/ing/emojify/app/model/ScriptFonts.kt
package ing.emojify.app.model

const val LATIN = "latin"

private val LANG_SCRIPT = mapOf(
    "ar" to "arabic",
    "bg" to "cyrillic",
    "el" to "greek",
    "he" to "hebrew",
    "hi" to "devanagari",
    "mr" to "devanagari",
    "ja" to "japanese",
    "ko" to "korean",
    "ru" to "cyrillic",
    "th" to "thai",
    "uk" to "cyrillic",
    "zh" to "chinese",
)

fun scriptForLang(lang: String?): String = LANG_SCRIPT[lang] ?: LATIN

private val HEBREW_RE = Regex("[\\u0590-\\u05FF]")

fun langForText(text: String): String = if (HEBREW_RE.containsMatchIn(text)) "he" else "en"

private val SCRIPT_FONTS: Map<String, Map<String, String>> = mapOf(
    "cyrillic" to mapOf(
        "anger" to "Russo One", "joy" to "M PLUS Rounded 1c", "play" to "Pacifico",
        "calm" to "Comfortaa", "tender" to "Marck Script", "drive" to "Unbounded",
        "sad" to "Cormorant Garamond", "anxiety" to "Amatic SC", "reflective" to "PT Serif",
    ),
    "greek" to mapOf(
        "anger" to "Tektur", "joy" to "Arima", "play" to "Mansalva",
        "calm" to "Comfortaa", "tender" to "Mynerve", "drive" to "Syne",
        "sad" to "Cardo", "anxiety" to "Fira Sans Extra Condensed", "reflective" to "Noto Sans",
    ),
    "arabic" to mapOf(
        "anger" to "Blaka", "joy" to "Marhey", "play" to "Lemonada",
        "calm" to "Mada", "tender" to "Playpen Sans Arabic", "drive" to "Alexandria",
        "sad" to "Amiri", "anxiety" to "Jomhuria", "reflective" to "Noto Sans Arabic",
    ),
    "hebrew" to mapOf(
        "anger" to "Suez One", "joy" to "Varela Round", "play" to "Gveret Levin",
        "calm" to "M PLUS Rounded 1c", "tender" to "Secular One", "drive" to "Miriam Libre",
        "sad" to "Frank Ruhl Libre", "anxiety" to "Amatic SC", "reflective" to "Heebo",
    ),
    "devanagari" to mapOf(
        "anger" to "Bakbak One", "joy" to "Modak", "play" to "Ranga",
        "calm" to "Palanquin", "tender" to "Kalam", "drive" to "Rajdhani",
        "sad" to "Tiro Devanagari Hindi", "anxiety" to "Khand", "reflective" to "Noto Sans Devanagari",
    ),
    "thai" to mapOf(
        "anger" to "Srisakdi", "joy" to "Chonburi", "play" to "Itim",
        "calm" to "Mitr", "tender" to "Charmonman", "drive" to "Chakra Petch",
        "sad" to "Taviraj", "anxiety" to "Sriracha", "reflective" to "Sarabun",
    ),
    "japanese" to mapOf(
        "anger" to "Dela Gothic One", "joy" to "Kosugi Maru", "play" to "Hachi Maru Pop",
        "calm" to "Zen Maru Gothic", "tender" to "Klee One", "drive" to "RocknRoll One",
        "sad" to "Shippori Mincho", "anxiety" to "DotGothic16", "reflective" to "Noto Sans JP",
    ),
    "korean" to mapOf(
        "anger" to "Black Han Sans", "joy" to "Jua", "play" to "Gaegu",
        "calm" to "Gowun Dodum", "tender" to "Nanum Pen Script", "drive" to "Do Hyeon",
        "sad" to "Nanum Myeongjo", "anxiety" to "Single Day", "reflective" to "Gothic A1",
    ),
    "chinese" to mapOf(
        "anger" to "WDXL Lubrifont SC", "joy" to "ZCOOL KuaiLe", "play" to "Liu Jian Mao Cao",
        "calm" to "ZCOOL XiaoWei", "tender" to "Zhi Mang Xing", "drive" to "ZCOOL QingKe HuangYou",
        "sad" to "Noto Serif SC", "anxiety" to "Long Cang", "reflective" to "Noto Sans SC",
    ),
)

private const val FALLBACK_FONT = "Noto Sans"

fun fontForScript(script: String, cluster: String): String = SCRIPT_FONTS[script]?.get(cluster) ?: FALLBACK_FONT
